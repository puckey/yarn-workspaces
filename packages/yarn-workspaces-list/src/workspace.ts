export type Name = string;
export type AbsolutePath = string;
export type RelativePath = string;

export interface Workspace {
  name: Name;
  path: AbsolutePath;
  location: RelativePath;
  workspaceDependencies: Name[];
  mismatchedWorkspaceDependencies: Name[];
  transitiveWorkspaceDependencies: Name[];
}
