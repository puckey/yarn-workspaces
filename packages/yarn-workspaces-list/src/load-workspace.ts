import { join } from 'path';
import { listWorkspacesByName } from './list';
import { pathExists, readJson } from './utils/fs';
import { Workspace } from './workspace';

export async function loadWorkspace(cwd: string): Promise<Workspace> {
  const pkgPath = join(cwd, 'package.json');
  if (!(await pathExists(pkgPath))) {
    throw new Error(`Could not find workspace at "${cwd}"`);
  }

  const { name } = await readJson(pkgPath);
  const byName = await listWorkspacesByName({ cwd });
  if (!byName[name]) {
    throw new Error(`"${name}" is not a workspace in the current project`);
  }

  return byName[name];
}
