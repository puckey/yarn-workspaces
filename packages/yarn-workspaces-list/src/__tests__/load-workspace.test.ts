import { join, relative } from 'path';
import { loadWorkspace } from '../load-workspace';
import { Workspace } from '../workspace';

jest.mock('child_process');

test('should load workspace', async () => {
  const fixtures = join(__dirname, '../../__fixtures__');
  const cwd = join(fixtures, 'packages/c');

  const workspace = await loadWorkspace(cwd);

  expect(normalize(cwd)(workspace)).toMatchSnapshot();
});

function normalize(cwd: string) {
  return (workspace: Workspace): Workspace => ({
    ...workspace,
    path: `absolute(${relative(cwd, workspace.path)})`,
  });
}
