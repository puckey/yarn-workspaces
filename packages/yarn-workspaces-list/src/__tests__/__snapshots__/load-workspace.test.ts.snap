// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should load workspace 1`] = `
Object {
  "location": "packages/c",
  "mismatchedWorkspaceDependencies": Array [],
  "name": "c",
  "path": "absolute()",
  "transitiveWorkspaceDependencies": Array [],
  "workspaceDependencies": Array [
    "a",
  ],
}
`;
